### PR TITLE
AArch64: Initial version of createPrologue/Epilogue() for SystemLinkage

### PR DESCRIPTION
**Pull request by knn-k**
This commit implements createPrologue() and createEpilog() for
ARM64SystemLinkage.

Signed-off-by: knn-k <konno@jp.ibm.com>

### CODE REVIEW

File: compiler/aarch64/codegen/ARM64SystemLinkage.cpp [hunk 1]
```diff
@@ -188,18 +330,116 @@ TR::ARM64SystemLinkage::createPrologue(TR::Instruction *cursor)
...
+   // save all callee-saved registers
+   // TODO: use the "stp" instruction when it becomes available
+   for (int r = TR::RealRegister::x19; r <= TR::RealRegister::x28; r++)
+      {
+      TR::MemoryReference *stackSlot = new (trHeapMemory()) TR::MemoryReference(sp, (r - TR::RealRegister::x19) * 8 + bodySymbol->getLocalMappingCursor() + 8, codeGen);
+      cursor = generateMemSrc1Instruction(cg(), TR::InstOpCode::strimmx, firstNode, stackSlot, machine->getARM64RealRegister((TR::RealRegister::RegNum)r), cursor);
+      }
+   for (int r = TR::RealRegister::v8; r <= TR::RealRegister::v15; r++)
+      {
+      TR::MemoryReference *stackSlot = new (trHeapMemory()) TR::MemoryReference(sp, (r - TR::RealRegister::v8) * 8 + (TR::RealRegister::x29 - TR::RealRegister::x19) * 8 + bodySymbol->getLocalMappingCursor() + 8, codeGen);
```
Comment (0xdaryl) [inline compiler/aarch64/codegen/ARM64SystemLinkage.cpp]: Can you break this line across multiple lines?  It's 221 characters long.

File: compiler/aarch64/codegen/ARM64SystemLinkage.cpp [hunk 2]
```diff
@@ -164,19 +166,159 @@ TR::ARM64SystemLinkage::getRightToLeft()
...
+               nextFltArgReg++;
+               mapSingleParameter(parameter, stackIndex);
+               }
+            else
+               {
+               nextFltArgReg = getProperties().getNumFloatArgRegs() + 1;
+               }
+            break;
+         case TR::Aggregate:
+            TR_ASSERT(false, "Function parameters of aggregate types are not currently supported on AArch64.");
```
Comment (0xdaryl) [inline compiler/aarch64/codegen/ARM64SystemLinkage.cpp]: Add a `default` case that asserts for the other possible data types.

File: compiler/aarch64/codegen/ARM64SystemLinkage.cpp [hunk 3]
```diff
@@ -164,19 +166,159 @@ TR::ARM64SystemLinkage::getRightToLeft()
...
+               {
+               nextFltArgReg++;
+               }
+            else
+               {
+               mapSingleParameter(parameter, stackIndex);
+               }
+            break;
+         case TR::Aggregate:
+            TR_ASSERT(false, "Function parameters of aggregate types are not currently supported on AArch64.");
```
Comment (0xdaryl) [inline compiler/aarch64/codegen/ARM64SystemLinkage.cpp]: Add a `default` case that asserts for the other possible data types.

File: compiler/aarch64/codegen/ARM64SystemLinkage.cpp [hunk 4]
```diff
@@ -188,18 +330,116 @@ TR::ARM64SystemLinkage::createPrologue(TR::Instruction *cursor)
...
+               cursor = generateMemSrc1Instruction(cg(), TR::InstOpCode::vstrimmd, firstNode, stackSlot, machine->getARM64RealRegister((TR::RealRegister::RegNum)(TR::RealRegister::v0 + nextFltArgReg)), cursor);
+               nextFltArgReg++;
+               }
+            else
+               {
+               nextFltArgReg = getProperties().getNumFloatArgRegs() + 1;
+               }
+            break;
+         case TR::Aggregate:
+            TR_ASSERT(false, "Function parameters of aggregate types are not currently supported on AArch64.");
```
Comment (0xdaryl) [inline compiler/aarch64/codegen/ARM64SystemLinkage.cpp]: Add a `default` case that asserts for the other possible data types.

File: compiler/aarch64/codegen/ARM64SystemLinkage.cpp [hunk 5]
```diff
@@ -188,18 +330,116 @@ TR::ARM64SystemLinkage::createPrologue(TR::Instruction *cursor)
...
+         }
+      }
+
+   // save all callee-saved registers
+   // TODO: use the "stp" instruction when it becomes available
+   for (int r = TR::RealRegister::x19; r <= TR::RealRegister::x28; r++)
```
[the remark this answers is not in the document]
Comment (knn-k) [reply compiler/aarch64/codegen/ARM64SystemLinkage.cpp]: The ARM version of `createPrologue/Epilogue()` does not use `getHasBeenAssignedInMethod()`, and I wrote the ARM64 version looking at it.  Rewriting using  `getHasBeenAssignedInMethod()`.

**Comment by knn-k** [comment]
Updated the file reflecting review comments.  The case with large stack frame is left as "not implemented yet".

**Pull request merged**